demos/time_dependent_rayleigh_benard/timedep-rayleigh-benard.py.rst

```diff
@@ -0,0 +1,174 @@
+and `Pablo Brubeck <https://www.maths.ox.ac.uk/people/pablo.brubeckmartinez/>`_.
+
+This problem involves a variable-temperature incompressible fluid.
+Variations in the fluid temperature are assumed to affect the momentum
+balance through a buoyant term (the Boussinesq approximation), leading
+to a Navier-Stokes equation with a nonlinear coupling to a
+convection-diffusion equation for temperature.
+
+We will set up the problem using Taylor-Hood elements for
+the Navier-Stokes part, and piecewise linear elements for the
+temperature. ::
+
+  from firedrake import *
+  from firedrake.pyplot import FunctionPlotter, tripcolor
+  import math
+  import matplotlib.pyplot as plt
+  from matplotlib.animation import FuncAnimation
+
+  try:
+      from irksome import Dt, RadauIIA, TimeStepper
+  except ImportError:
+      import sys
+      warning("Unable to import irksome.")
+      sys.exit(0)
+
+  Nbase = 8
+  ref_levels = 2
+  N = Nbase * 2**ref_levels
+
+  base_msh = UnitSquareMesh(Nbase, Nbase)
+  mh = MeshHierarchy(base_msh, ref_levels)
+  msh = mh[-1]
+
+  V = VectorFunctionSpace(msh, "CG", 2)
+  W = FunctionSpace(msh, "CG", 1)
+  Q = FunctionSpace(msh, "CG", 1)
+  Z = V * W * Q
+
+  upT = Function(Z)
+  u, p, T = split(upT)
+  v, q, S = TestFunctions(Z)
+
+Two key physical parameters are the Rayleigh number (Ra), which
+measures the ratio of energy from buoyant forces to viscous
+dissipation and heat conduction and the
+Prandtl number (Pr), which measures the ratio of viscosity to heat
+conduction. ::
+
+  Ra = Constant(2000.0)
+  Pr = Constant(6.8)
+
+Along with gravity, which points down. ::
+
+  g = Constant((0, -1))
+
+Set up variables for time and time-step size. ::
+
+  t = Constant(0.0)
+  dt = Constant(1.0 / N)
+
+  F = (
+      inner(Dt(u), v)*dx
+      + inner(grad(u), grad(v))*dx
+      + inner(dot(grad(u), u), v)*dx
+      - inner(p, div(v))*dx
+      - (Ra/Pr)*inner(T*g, v)*dx
+      + inner(div(u), q)*dx
+      + inner(Dt(T), S)*dx
+      + inner(dot(grad(T), u), S)*dx
+      + 1/Pr * inner(grad(T), grad(S))*dx
+  )
+
+There are two common versions of this problem.  In one case, heat is
+applied from bottom to top so that the temperature gradient is
+enforced parallel to the gravitation.  In this case, the temperature
+difference is applied horizontally, perpendicular to gravity.  It
+tends to make prettier pictures for low Rayleigh numbers, but also
+tends to take more Newton iterations since the coupling terms in the
+Jacobian are a bit stronger.  Switching to the first case would be a
+simple change of bits of the boundary associated with the second and
+third boundary conditions below::
+
+  bcs = [
+      DirichletBC(Z.sub(0), Constant((0, 0)), (1, 2, 3, 4)),
+      DirichletBC(Z.sub(2), Constant(1.0), (3,)),
+      DirichletBC(Z.sub(2), Constant(0.0), (4,))
+  ]
+
+Like Navier-Stokes, the pressure is only defined up to a constant.::
+
+  nullspace = [(1, VectorSpaceBasis(constant=True))]
```

Comment:
This looks wrong to me - we have 3 spaces in `Z`, but only 2 in the `nullspace`...

Reply:
Good catch.  It was running, but I fixed it.